File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys
input = sys.stdin.readline
n = int(input()) #전체사람수
a,b = map(int,input().split()) # 두사람의 촌수를 알아야함
m = int(input()) #부모 자식들 간의 관계수 즉,,,약간 노드...?
graph = [[] for _ in range(n+1)]
visited = [False]*(n+1)
result = []
for _ in range(m):
    x, y = map(int,input().split())
    graph[x].append(y)
    graph[y].append(x)

def dfs(v,cnt):
    cnt +=1
    visited[v] = True
    if v== b:
        result.append(cnt)
    for i in graph[v]:
        if not visited[i]:
            dfs(i,cnt)
dfs(a,0)

if len(result)==0:
    print(-1)
else:
    print(result[0]-1)

File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.*;

import org.xml.sax.SAXException;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int a = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(br.readLine());
        
        //그래프 초기화
        List<List<Integer>> graph = new ArrayList<>();
        for(int i=0;i<=n;i++){
            graph.add(new ArrayList<>());
        }

        boolean[]visited = new boolean[n+1];
        List<Integer> result = new ArrayList<>();
        

        for(int i=0;i<m;i++){
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            
            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        dfs(graph, visited,result, a,b,0);
        if(result.isEmpty()){
            System.out.println(-1);
        }else{
            System.out.print(result.get(0)-1);
        }
    }
    static void dfs(List<List<Integer>> graph, boolean[] visited,List<Integer> result, int v, int target, int cnt){
        cnt +=1;
        visited[v] = true;
        if(v==target){
            result.add(cnt);
            return;
        }
        for(int next:graph.get(v)){
            if(!visited[next]){
                dfs(graph, visited, result, next, target,cnt);
            }
        }
    }
}